src/Microsoft.IdentityModel.Tokens/EventBasedLRUCache.cs

```diff
@@ -534,7 +590,7 @@ public bool TryRemove(TKey key, out TValue value)
             }
 
             value = cacheItem.Value;
-            OnItemRemoved?.Invoke(cacheItem.Value);
+            OnItemCompacted?.Invoke(cacheItem.Value);
```

Comment:
In both of these TryRemove methods, shouldn't we add the cacheItem to the `_compactedItems` like we do in Compact methods?

Because currently, if TryRemove is called, OnItemCompacted will just set the `signatureProvider.CryptoProviderCache = null` but since the provider is never added to `_compactedItems`, OnItemRemoved will never be called for it, which disposes of it.

Reply:
In our call graph the SignatureProvider calls this when the SignatureProvider was disposed or Signing / Verifying faulted. 
In the case where the finally clause is executed, dispose is called on the cached objects.

A better design would be to have TryRemove move the SignatureProvider to the compacted list.
This would need a new action in LRUCache.

For now i think it is OK but a bit fragile.